src/compiler/checker.ts

```diff
@@ -24323,18 +24323,30 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
      * and no required properties, call/construct signatures or index signatures
      */
     function isWeakType(type: Type): boolean {
-        if (type.flags & TypeFlags.Object) {
-            const resolved = resolveStructuredTypeMembers(type as ObjectType);
-            return resolved.callSignatures.length === 0 && resolved.constructSignatures.length === 0 && resolved.indexInfos.length === 0 &&
-                resolved.properties.length > 0 && every(resolved.properties, p => !!(p.flags & SymbolFlags.Optional));
-        }
-        if (type.flags & TypeFlags.Substitution) {
-            return isWeakType((type as SubstitutionType).baseType);
-        }
-        if (type.flags & TypeFlags.Intersection) {
-            return every((type as IntersectionType).types, isWeakType);
+        return isWeakTypeWorker(type) > 0;
+
+        function isWeakTypeWorker(type: Type): -1 | 0 | 1 {
+            if (type.flags & TypeFlags.Object) {
+                const resolved = resolveStructuredTypeMembers(type as ObjectType);
+                return resolved.callSignatures.length === 0 && resolved.constructSignatures.length === 0 && resolved.indexInfos.length === 0 &&
+                        resolved.properties.length > 0 ? (every(resolved.properties, p => !!(p.flags & SymbolFlags.Optional)) ? 1 : 0) : -1;
+            }
+            if (type.flags & TypeFlags.Substitution) {
+                return isWeakTypeWorker((type as SubstitutionType).baseType);
+            }
+            if (type.flags & TypeFlags.Intersection) {
+                let result: -1 | 0 | 1 = -1;
+                for (const t of (type as IntersectionType).types) {
+                    const v = isWeakTypeWorker(t);
+                    if (!v) {
+                        return 0;
+                    }
+                    result = result > v ? result : v;

```

Comment:
The idea behind the fix is that an empty object type should not be considered to be weak on its own. That's comes from the pre-existing requirement for weak types:
```ts
resolved.properties.length > 0 && every(resolved.properties, p => !!(p.flags & SymbolFlags.Optional))
```

However, when it's intersected with a weak type that should not make it, out of a sudden, non-weak. The new logic still maintains the invariant that an intersection of many empty interfaces  is not considered to be weak (it should obey the same rules as a single empty interface). 

All of that doesn't change the definition put in the existing code comment:
```ts
    /**
     * A type is 'weak' if it is an object type with at least one optional property
     * and no required properties, call/construct signatures or index signatures
     */
```

Actually, the old behavior wasn't enforcing those rules for intersections precisely enough. The proposed adjustment is closer to checking the intersection as a single type with a flat list of properties instead of checking separate types and combining those results "blindly". I think it's more accurate to do it this way because for most purposes an intersection shouldn't be distinguishable from the same type with a "flattened" list of properties.